### ISSUE TITLE
map! prevents side effects in nom 5

### ISSUE THREAD

**Issue by ankoh** [Geal/nom#949]
Hey all,

I ran into a quite far-reaching issue while migrating to the nom 5 beta.
If I got the diff correctly, you introduced a tiny function call for the map! macro here:
https://github.com/Geal/nom/commit/e6f3c6fcbe23bd456bf5474ead6c249a2b75188e

The argument, however, is a `Fn` closure which prevents mutable borrows.
Is this intended?
I started to use map! for side effects quite often.
I'm regularly scanning over a buffer to update an existing state in-flight via these side effects, effectively leaving the remaining state untouched.

That turned out to be a quite elegant pattern that is no longer working without interior mutability.
Can this argument be lifted to `FnMut`?

- Rust version : rustc 1.35.0 (3c235d560 2019-05-20)
- nom version : 5.0.0-beta1
- nom compilation features used: default

## Test case

Example test case:

```
#[macro_use]
extern crate nom;

use nom::number::complete::{le_u8, le_u32, le_f64};
use nom::IResult;

fn foo() {
  let buffer = &b"a"[..];
  let some_side_effect: f64 = 0.0;
  
  do_parse!(
    buffer,
    fields: le_u32 >>
    count!(|input: &[u8]| -> IResult<&[u8], ()> {
        switch!(input, le_u8,
            // No longer allowed in nom 5 due to Fn closure
            0 => map!(le_f64, |v| { some_side_effect = v })
        )},
        fields as usize
    ) >>
    (())
  );
}
```

**Comment by ankoh** [comment]
After skimming the issues I understand now that this function call is, in fact, a crucial aspect of nom 5. And immutable captures are probably sufficient most of the time.
But I found nom to be a very appealing tool to scan quite primitive wire formats.
And there I oftentimes do not really want to pay for a Vec<UmbrellaEnum> if a side effect per match does the same.

**Comment by Geal** [comment]
yes, I'm not sure switching to `FnMut` will be possible. It's still possible to get the side effect you want in a `Fn` with interior mutability, though. cf https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=4eacbb3f86ff67698e6b7596b42b7623

**Comment by ankoh** [comment]
I understand. Yes, that's exactly what I did first but it spread far to many Cells. I ended up folding state copies which unfortunately puts a fair amount of trust in the compiler.

I'd argue that side effects are probably not too uncommon especially in more performance sensitive scenarios.
If you have a rather small amount of output types and can live with dirty state, side effects will always be exactly what you want.
Otherwise the compiler has to do non-trivial reasoning to eliminate all the copies.
(Which is particularly undesirable if you use nom where it really shines - primitive recursive descent wire format decoders)
